uaparser/parser.go

```diff
@@ -384,33 +417,54 @@ func (parser *Parser) ParseDevice(line string) *Device {
 }
 
 func checkAndSort(parser *Parser) {
-	parser.Lock()
+	parser.mu.Lock()
 	if atomic.LoadUint64(&parser.UserAgentMisses) >= parser.config.MissesThreshold {
 		if parser.config.DebugMode {
 			fmt.Printf("%s\tSorting UserAgents slice\n", time.Now())
 		}
 		parser.UserAgentMisses = 0
-		sort.Sort(UserAgentSorter(parser.UA))
+
+		// create a copy to avoid modifying the original slice while in use.
+		uaDefinitions := make([]*uaParser, len(parser.RegexDefinitions.UA))
+		copy(uaDefinitions, parser.RegexDefinitions.UA)
+
+		sort.Sort(UserAgentSorter(uaDefinitions))
+
+		parser.RegexDefinitions.UA = uaDefinitions
```

Comment:
```suggestion
		sort.Sort(UserAgentSorter(parser.RegexDefinitions.UA))
```
since we're now cloning on construction in the UseSort case, and this is protected by a lock - probably we can go back to sorting in place rather than doing a copy here? is there any advantage to copying?